code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,364 @@
+            unit.start();
+        }
+    }
+
+    protected List<DeploymentUnit> deleteBadUnits(List<DeploymentUnit> units) {
+        List<DeploymentUnit> result = new ArrayList<>(units.size());
+
+        for (DeploymentUnit unit : units) {
+            if (unit.isError()) {
+                unit.remove();
+            } else {
+                result.add(unit);
+            }
+        }
+
+        return result;
+    }
+
+    protected List<DeploymentUnit> matchScale(List<Service> services, List<DeploymentUnit> units, int requestedScale) {
+        /*
+         * If there are too many units, call unit.remove() and delete the excess.
+         * If there are not enough units, create new empty deployment units by call the factory
+         * and passing in the services and an empty list for instances.
+         * 
+         * NOTE: We don't actually start services hereed
+         */
+
+        if (units.size() < requestedScale) {
+            addMissingUnits(services, units, requestedScale);
+        } else if (units.size() > requestedScale) {
+            removeExtraUnits(units, requestedScale);
+        }
+
+        return units;
+    }
+
+    private void removeExtraUnits(List<DeploymentUnit> units, int scale) {
+        // delete units
+        int i = units.size() - 1;
+        while (units.size() > scale) {
+            DeploymentUnit toRemove = units.get(i);
+            toRemove.remove();
+            units.remove(i);
+            i--;
+        }
+    }
+
+    private void addMissingUnits(List<Service> services, List<DeploymentUnit> units, int scale) {
+        Map<Service, List<String>> availableNames = new HashMap<>();
+        // add more units
+        for (Service service : services) {
+            List<String> instanceNames = sdService.getAvailableServiceInstanceNames(service,
+                    scale - units.size());
+            availableNames.put(service, instanceNames);
+        }
+        
+        int i = 0;
+        while (units.size() < scale) {
+            String uuid = UUID.randomUUID().toString();
+            Map<Service, DeploymentUnitInstance> serviceToInstances = new HashMap<>();
+            for (Service service : services) {
+                Pair<String, Object> instanceNamePair = Pair.of(availableNames.get(service).get(i),
+                        null);
+                DeploymentUnitInstance deploymentUnitInstance = deploymentUnitInstanceFactory
+                        .createDeploymentUnitInstance(uuid, service, instanceNamePair);
+                serviceToInstances.put(service, deploymentUnitInstance);
+            }
+            DeploymentUnit unit = new DeploymentUnit(uuid, serviceToInstances);
+
+            units.add(unit);
+            i++;
+        }
+    }
+
+    private void fillUpMissingUnitInstances(List<Service> services, List<DeploymentUnit> units) {
+        // check if there are incomplete units
+        Map<Service, Integer> serviceToUnitInstanceCount = new HashMap<>();
```

Comment:
I do it outside - in the manager - only because I'm trying to preserve order for the instances names. If several units have to fill up the gaps for several services, for each service I grab the available names (and fill up the gaps in names if needed), then redistribute the names across units.